### ISSUE TITLE
Charts is not fit to parent until screen is resized

### ISSUE THREAD

**Issue by HyunnoH** [apexcharts/react-apexcharts#187]
Hi, I am using react-apexcharts and set height as fixed 300px. I thought the chart's width should be set typed or fit to parent tag. But when I using react-apexcharts, it overs parent div.

Here is my chart code:
```typescript
export default function ProjectStatusGraph() {
  const options = {
    chart: {
      toolbar: {
        show: false,
      },
      id: 'project-status-graph',
    },
    dataLabels: {
      enabled: false,
    },
    stroke: {
      curve: 'smooth',
    },
    xaxis: {
      type: 'datetime',
      labels: {
        show: true,
        showDuplicates: true,
        datetimeUTC: true,
      },
    },
  };

  const activatedProjectMock = projectUsageMock
    .map(r => [Number(r.date), r.projects.filter(p => p.activated).length])
    .sort((a, b) => a[0] - b[0]);
  const workingProjectMock = projectUsageMock
    .map(r => [Number(r.date), r.projects.filter(p => p.activated && p.usage > 10).length])
    .sort((a, b) => a[0] - b[0]);

  return (
    <Card title="Project Activated / Working" size="small">
      <ReactApexChart
        options={options}
        series={[
          {
            name: 'activated',
            data: activatedProjectMock,
          },
          {
            name: 'working',
            data: workingProjectMock,
          },
        ]}
        height={300}
        type="area"
      />
    </Card>
  );
}
```

And here is behavior: 
![ezgif com-video-to-gif](https://user-images.githubusercontent.com/43626722/84632915-99ad9a80-af2a-11ea-8992-cdd077eed648.gif)

The chart is overflowed until I adjust screen width or some re-rendering. Is it a bug or correct behavior of apexcharts which can handled by developer?

**Comment by ckienast** [comment]
I have a similar problem. I have two charts on one page and if the page gets resized, the charts get bigger, but not small again.

This is my implementation: https://gist.github.com/ckienast/261b919257dd598afc1aa27b5f41277b

**Comment by HyunnoH** [comment]
I fixed this by redeclaring apex object. I seems apexcharts needs re-rendering to find and get parent tag's size.

[here](https://gist.github.com/HyunnoH/bdad94c9ad2d9a5c9b2a4660737a12c6) is my gist about how I avoided this issues.

**Comment by kkirsche** [comment]
I'm also experiencing this when a sidebar is triggered in and out. The flow, via images, is:

Sidebar starts off opened:
<img width="1184" alt="Screen Shot 2020-06-19 at 05 58 27" src="https://user-images.githubusercontent.com/947110/85120945-f68ea680-b1f1-11ea-8943-03db6a622bed.png">

Sidebar is closed: (note the shift)
<img width="1288" alt="Screen Shot 2020-06-19 at 05 58 34" src="https://user-images.githubusercontent.com/947110/85120953-f9899700-b1f1-11ea-8329-54f1c869482c.png">

Sidebar is re-opened:
<img width="1192" alt="Screen Shot 2020-06-19 at 05 58 40" src="https://user-images.githubusercontent.com/947110/85120960-fbebf100-b1f1-11ea-8bf8-4b65a001b2de.png">

Implementation:
https://gist.github.com/kkirsche/3aa5d3a790fbee627d3cca1f5454b3f8

Any tips on how to fix this would be greatly appreciated

**Comment by junedchhipa** [comment]
Please specify the version of `apexcharts` as well as `react-apexcharts` as this was an old issue already fixed.

**Comment by ckienast** [comment]
`"apexcharts": "^3.19.2"`
`"react-apexcharts": "^1.3.7"`

**Comment by kkirsche** [comment]
same versions on my side as @ckienast

**Comment by francoiscabrol** [comment]
I also have a similar issue with the last version when I set the height as props. It is described in the issue #180

**Comment by oscarcornejo** [comment]
Hi, I have the same problem. Any solution?  the issue[ #180](https://github.com/apexcharts/react-apexcharts/issues/180) that @francoiscabrol comments blocks the animations.

**Comment by JalmariO** [comment]
I have the same problem when using angular material and sidenav. Latest Apexcharts from npm.

**Comment by philipp-schmidt** [comment]
Same behavior here. Charts will not fit into parent container until the first update of their data. Any ideas? I tried loading the data "delayed" by loading them inside `useEffect()`, but it did not help. I also tried to invoke `window.dispatchEvent(new Event('resize'));` but that did not chance anything either.

**Comment by simplecommerce** [comment]
Hi, I have the same problem but it seems random, here I have 2 charts, one can be resized the other, doesn't resize at all.

https://codesandbox.io/s/grid-layout-react-apexcharts-2jdo2?file=/src/App.js

In the side window it does work but if you click on (Open in new window) and resize, it won't work on the 2nd chart.

If I open the dev tools and resize the whole window, then both charts work.

I am using latest version of both Apexcharts and react-apexcharts.

Is there a fix for this?

**Comment by neilgrewal** [comment]
I have the same issue using the angular library. The solution I'm currently using is the following:

sidebar.service.ts
```
toggleSubject: Subject<boolean> = new Subject<boolean>();

toggle(): void {
   ...
   const togggled = ...
   this.toggleSubject.next(toggled);
}

getToggleSubject(): Observable<boolean> {
    return this.toggleSubject.asObservable();
  }
```
component.ts
```
this.sidebarToggleSubscriber = this.sidebarService.getToggleSubject().subscribe(() => {
      this.render();
    });

render(): void {
    setTimeout(() => {
        window.dispatchEvent(new Event('resize'));
      }, 300);
  }
```

I use 300 since thats the transition time my sidenav takes to appear/disappear.

**Comment by Pavinati** [comment]
I have the same problem. In the meantime I used the `mounted` event as a work-around the issue.

```
  const chartOptions = {
    chart: {
      events: {
        mounted: (chart) => {
          chart.windowResizeHandler();
        }
      }
    }
  };
```

**Comment by darkcurrent** [comment]
> I have the same problem. In the meantime I used the `mounted` event as a work-around the issue.
> 
> ```
>   const chartOptions = {
>     chart: {
>       events: {
>         mounted: (chart) => {
>           chart.windowResizeHandler();
>         }
>       }
>     }
>   };
> ```

This entry solved the issue for me.

**Comment by Tergeltengis** [comment]
Also this happens my chart  type='bar' and I'm add this in my code.
  
plotOptions: {
      bar: { horizontal: false, barHeight: '28%', borderRadius: 2 }
    },

**Comment by charliearaya** [comment]
I see this issue when I have multiple charts and the parent div has the following css properties:
`display: grid;
grid-template-columns: 1fr 1fr;
`

Currently using versions:
"apexcharts": "^3.33.0",
"react-apexcharts": "^1.3.9",

**Comment by AntonyOnScript** [comment]
> > I have the same problem. In the meantime I used the `mounted` event as a work-around the issue.
> > ```
> >   const chartOptions = {
> >     chart: {
> >       events: {
> >         mounted: (chart) => {
> >           chart.windowResizeHandler();
> >         }
> >       }
> >     }
> >   };
> > ```
> 
> This entry solved the issue for me.

thanksss!

**Comment by mennovanhout** [comment]
> I have the same problem. In the meantime I used the `mounted` event as a work-around the issue.
> 
> ```
>   const chartOptions = {
>     chart: {
>       events: {
>         mounted: (chart) => {
>           chart.windowResizeHandler();
>         }
>       }
>     }
>   };
> ```

This is a nice solution but really messes up the animation when loading in.
Is there another fix found?

**Comment by georgesunnyt** [comment]
I'm also currently facing this issue.

**Comment by enesbasar** [comment]
> > I have the same problem. In the meantime I used the `mounted` event as a work-around the issue.
> > ```
> >   const chartOptions = {
> >     chart: {
> >       events: {
> >         mounted: (chart) => {
> >           chart.windowResizeHandler();
> >         }
> >       }
> >     }
> >   };
> > ```
> 
> This is a nice solution but really messes up the animation when loading in. Is there another fix found?

I used "animationEnd" event instead of "mounted". Seems to work fine.

**Comment by AniaKru95** [comment]
I had the same problem when my sidebar has been expanded, chart was hidden outside window.
That's weird but I've added min-width: 0 to parent div and chart resizes right now.

```
<div style="min-width:0">
      <apexchart />
</div>
```

**Comment by shazyQ** [comment]
Hi, I have the same problem and I used the mounted event but its working only in chrome browser not in mozilla firefox.

```
var options = {
        colors: [function ({ value, seriesIndex, w }) {
            if (value >= 0) {
                return '#feb600'
            } else {
                return '#c1c3c1'
            }
        }, function ({ value, seriesIndex, w }) {
            if (value >= 0) {
                return '#3ca2db'
            } else {
                return '#415364'
            }
        }],
        series: [{
            name: 'Previous Period',
            data: previousYearDataByTargetLang
        }, {
            name: 'Current Period',
            data: currentYearDataByTargetLang
            }],
        chart: {
            type: 'bar',
            height: xAxisTargetLang.length <= 8 ? 430 : xAxisTargetLang.length * 60
            , toolbar: {
                show: false
            }, events: {
                mounted: (chart) => {
                    chart.windowResizeHandler();
                }
            },
            fontFamily: 'Noto Sans, sans- serif'
        },
        plotOptions: {
            bar: {
                horizontal: true,
                dataLabels: {
                    position: 'top',
                },
            }
        },
        dataLabels: {
            enabled: false
        },
        stroke: {
            show: true,
            width: 1,
            colors: ['#fff']
        },
        tooltip: {
            shared: true,
            intersect: false,
            y: {
                formatter: function (value, { series, seriesIndex, dataPointIndex, w }) {
                    return "$" + addThousandSeperator(Math.round(value))
                }
            }
        },
        xaxis: {
            labels: {
                formatter: function (val, index) {
                    return addThousandSeperator(val.toFixed(0));
                }
            },
            categories: xAxisTargetLang,
        }

    };   
    var chart = new ApexCharts(document.querySelector("#TargetLanguageChart"), options);
    chart.render();
```

https://user-images.githubusercontent.com/117825585/200823025-8c766617-f6b7-4b7d-8f79-51bf13cdd18a.mp4

**Comment by AmarokStudios** [comment]
I am having the same issue with version V3.31.0. A quick workaround that has been working for me so far is to use jQuery's document ready function and resizing the window with the event dispatcher. It's not the most ideal, but it works for me.

```js
$( document ).ready(function() {
    window.dispatchEvent(new Event("resize"));
});
```

**Comment by Haseeb-Ahmed-HA** [comment]
I faced the same issue: when i close side bar the chart not changing its width and when sidebar is open chart width overflows to parent. On resizing window it works fine, but if we don't resize window it only changes its width according to parent when rerender. Here is the bug. I am using react and I fixed this issue by rerender chart componet when sidebar is open or close.

  useEffect(() => {
    const timeOut = setTimeout(() => {
      setRender(isOpen);
    }, 500);
    return () => clearTimeout(timeOut);
  }, [isOpen]);

now it works fine you can simply add the dependency in useEffect when you want to rerender chart component.

**Comment by Zoot01** [comment]
"apexcharts": "^3.37.0"

Still in issue in this latest version. Sidebar will randomly cause resizing issues with my "line" chart. current resolution seems to be utilizing the useEffect hook to force rerender. 

```typescript
  useEffect(() => {
    const timeOut = setTimeout(() => {
      const chartInstance = ApexCharts.getChartByID("visits-chart")
      if (chartInstance) chartInstance.updateSeries(series, true)
    }, 500)
    return () => clearTimeout(timeOut)
  }, [sidebarIsOpen])
```

chartInstance.updateSeries() takes two parameters, the chart data, and a boolean value forcing a new animation. Hope this helps. By the way this issue only seems to occur with the line chart. Let me know if anyone else is still experiencing this issue.

**Comment by Coderamrin** [comment]
> > > I have the same problem. In the meantime I used the `mounted` event as a work-around the issue.
> > > ```
> > >   const chartOptions = {
> > >     chart: {
> > >       events: {
> > >         mounted: (chart) => {
> > >           chart.windowResizeHandler();
> > >         }
> > >       }
> > >     }
> > >   };
> > > ```
> > 
> > 
> > This is a nice solution but really messes up the animation when loading in. Is there another fix found?
> 
> I used "animationEnd" event instead of "mounted". Seems to work fine. Edit: This one resizes after animation is done, so it might not be the best solution.

yeah, it add this weird effect after the animation. need the the animation and width work together.

**Comment by Coderamrin** [comment]
Is there any solution for this?

**Comment by simplecommerce** [comment]
> Is there any solution for this?

don't know if it helps but try adding a `setTimeout` when calling the `chart.windowResizeHandler();` of say `800` after the `mounted` is called.

**Comment by Coderamrin** [comment]
> > Is there any solution for this?
> 
> don't know if it helps but try adding a `setTimeout` when calling the `chart.windowResizeHandler();` of say `800` after the `mounted` is called.

Will try it out. 

A quick note here: I just discovered the horizon UI dashboard template uses apexchart in next js and it seems to work smoothly. 
currently inspecting their code. Will let you guys know if i find any fix. 


here's the template if anyone intersted: https://github.com/horizon-ui/horizon-ui-chakra-nextjs

**Comment by kaio-eduardo** [comment]
i have the same problem, for some reason some of the charts resize some dont, for now what i am doing is that:
` 
useEffect(() => {
    const timeout = setTimeout(() => {
      window.dispatchEvent(new Event('resize'));
    }, 500)
    
    return () => {
      clearTimeout(timeout)      
    }
  }, [open])
`
it works, but is there a better solution?

**Comment by Coderamrin** [comment]
can you please share the context where you added this code snippet?

**Comment by shazyQ** [comment]
Please find the snippets below, hope that will help.

 chart: {
            type: 'bar',
            height: 430,
            **events: {
                mounted: (chart) => {
                    chart.windowResizeHandler();
                }
            },**
        },

**Comment by kaio-eduardo** [comment]
> can you please share the context where you added this code snippet?

i have a sidebar that when i open the width of the content change, when that happens the charts dont resize, just overflow the parent div.

![image](https://github.com/apexcharts/react-apexcharts/assets/44307221/deb0e6c9-8d84-416a-bfda-aead8aba7760)

But every time a resize event happens, the width is adjusted, so I made it so that when the sidebar opens this event occurs and arranges the charts.

this problem only happens when there is data in the chart.

**Comment by VaibhavApplore** [comment]
@ginuxone did you tried my solution?

**Comment by tpalaz** [comment]
This is also happening for me. What's very frustrating is this only happens when more than 1 chart is rendered on the page. If there's a single chart, window sizing works as expected and the chart responds to the containers size. However if more than 1 chart is added (included an identical chart), the weird resizing / lack of responsiveness bug happens. Extremely frustrating to deal with.

**Comment by ginuxone** [comment]
@VaibhavApplore Yes, it works! I was wondering if this was the only solution for now. I have performance issues and this takes it's small toll

**Comment by mtuzenko** [comment]
Just had same issue using Angular "ng-apexcharts": "~1.8.0". The solution was to set `display: block` to `apx-chart`

**Comment by matheusjordan** [comment]
> Just had same issue using Angular "ng-apexcharts": "~1.8.0". The solution was to set `display: block` to `apx-chart`

this worked for me, thanks

**Comment by JalmariO** [comment]
> Just had same issue using Angular "ng-apexcharts": "~1.8.0". The solution was to set `display: block` to `apx-chart`

Yes! This works, thank you!

**Comment by Avladd** [comment]
The solutions suggested here did not work for me. 

What did was using a [ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver) to get the parent's width, and then passing that to the chart manually. This does come with a performance hit, but for my purposes it works well enough, and it also kept the animations.

```
  useLayoutEffect(() => {
    const handleChange: ResizeObserverCallback = entries => {
      for (const entry of entries) {
        if (entry.contentRect) {
          const graphWidth = entry.contentRect.width
          setReponsiveGraphWidth(graphWidth)
        }
      }
    }
    const el = document.getElementById('chart-wrapper')
    const observer = new ResizeObserver(handleChange)
    if (el) {
      observer.observe(el)
    }
    return () => {
      observer.disconnect()
    }
  })
  ```
  
  then I pass in the width from state
  
  ```
  <ReactApexChart
      type="bar"
      options={options}
      series={data as ApexAxisChartSeries}
      width={responsiveGraphWidth}
   ```

**Comment by lucasrachid** [comment]
> Just had same issue using Angular "ng-apexcharts": "~1.8.0". The solution was to set `display: block` to `apx-chart`

I work with "ng-apexcharts": "^1.11.0", project in Angular, and this is the unique solution that solved my problem. Im logged at git hub, just to thank you dude. Lost about 2-3 hours xD

**Comment by Vickyrdz** [comment]
I have the same problem using apexchart from koolreport, the graph does not adapt to the change in screen size, only if I refresh the entire screen and it is not what I am looking for, I want the graph to update and adapt when the screen changes. Does anyone know the solution?

**Comment by Fenharel1** [comment]
> Just had same issue using Angular "ng-apexcharts": "~1.8.0". The solution was to set `display: block` to `apx-chart`

This worked for me, my current version of apexcharts is "ng-apexcharts": "^1.7.4" in angular 18
